README.md
# flume-interceptor-ex

[![Build Status](https://travis-ci.org/supermy/rule-interceptor.svg?branch=master)](https://github.com/supermy/rule-interceptor)

##介绍

可以数据采集端，利用采集设备的能力对数据进行非业务类 进行实时的 ETL 处理；适用于 hadoop 大数据生态的各种通道环境；特色是采用 groovy 作为规则语言，能够完成各种条件的处理。

flume 全能版本拦截器 [flume-rule-interceptor](https://github.com/supermy/rule-interceptor). groovy 作为规则dsl,通过groovy 脚本编写逻辑，支持JSON 格式函数，脚本动态加载，灰度发布。

业务场景1：RuleSearchAndReplaceInterceptor， 互联网算进行数据传输的安全，通过拦截器进行加密解密；官方原有的正则不能实现此功能。

业务场景2：RuleFilteringInterceptor， 数据条件过滤，可以通过groovy 脚本进行条件过滤，非常灵活；官方原有的正则不支持条件。

业务场景3：RuleSearchAndReplaceInterceptor，数据格式变更，可以通过groovy 脚本进行数据格式转换；官方可以通过正则完成，效率较低。

业务场景3：RuleSearchAndReplaceInterceptor，定制head 属性，可以通过groovy 脚本配置head 属性；官方配置较为复杂，不能支持灵活业务。

[new]业务场景4：支持内容过滤；支持将内容转换为 redis-lua 支持的脚本，支持 flume-redis 组件进行数据处理。

## Install

安装:

    打包rule-interceptor-1.0-SNAPSHOT.jar拷贝到flume的plugins.d/flume-interceptor/lib目录
    拷贝groovy-all-2.4.7.jar到flume的plugins.d/flume-interceptor/libext目录

## Usage

    
    a1.sources.r1.interceptors.i2.type = com.supermy.flume.interceptor.RuleFilteringInterceptor$Builder
    a1.sources.r1.interceptors.i2.rule = /etc/flume/conf/g-filter.groovy
    a1.sources.r1.interceptors.i2.ruleName = filterGroovy
    
    a1.sources.r1.interceptors.i3.type = com.supermy.flume.interceptor.RuleSearchAndReplaceInterceptor$Builder
    a1.sources.r1.interceptors.i3.searchReplaceDsl = /etc/flume/conf/g-search-replace.groovy
    a1.sources.r1.interceptors.i3.searchReplaceKey = searchReplaceGroovy
    
        

### g-filter.groovy

过滤脚本，可以使用head and body 的数据作为条件 判定词条数据是否过滤

```  groovy

    //#当它的值为true 的时候，过滤掉匹配到当前正则表达式的一行
    //#当它的值为false的时候，就接受匹配到正则表达式的一行
    println "netuser filter"
    //println head
    //println body
    body = "20170621162925,113.225.23.151,test_10056368,1"
    //body = "20170621162925,113.225.23.152,test_10056368,3"
    
    def split = body.split(",")
    
    println split.size()
    
    if(split.size()<4){
        //数据不合格过滤
        return false;
    }
    
    def type = split[3]
    
    println type.getClass()
    println type.substring(0,1) == '1'
    //println type.toInteger() == 1
    
    if (type.substring(0,1) == '1' || type.substring(0,1) == '2') {
        return true;  //不过滤
    } else {
        return false; //过滤
    }

```



### g-search-replace.groovy

替换脚本，可以更改head and body 的数据，适配不同的业务场景，脚本支持动态更新；

``` groovy 加密场景

        import  com.supermy.flume.interceptor.*
        import javax.crypto.Cipher;
        import javax.crypto.spec.SecretKeySpec;
        
        println head
        println body
        body=body.replace('a','aaa')
        head["newhead"]='abcd'
        
        
        
        String text = "Body 的数据 , I Love BONC"
        
        //
        def key = new SecretKeySpec("123456789987654321".bytes, "AES")
        def c = Cipher.getInstance("AES")
        
        //加密
        c.init(Cipher.ENCRYPT_MODE, key)
        e_text = new String(Hex.encodeHex(c.doFinal(text.getBytes("UTF-8"))))
        
        //解密
        c.init(Cipher.DECRYPT_MODE, key)
        text1 = new String(c.doFinal(Hex.decodeHex(e_text.toCharArray())))
        
        println text
        println e_text
        println text1
        
        
        def resultMap = [:]
        
        //加密数据，用于互联网数据传输
        
        
        resultMap["head"]=head
        resultMap["body"]=body
        
        return resultMap

```

``` groovy redis-lua 场景

    import com.google.gson.Gson
    
    //数据加工为 lua 脚本，插入到 redis
    println "netuser redis script 准备"
    //println head
    //println body
    
    //body = "20170621162925,113.225.23.151,test_10056368,1"
    body = "20170621162925,113.225.23.152,test_10056368,2"
    // eval "return redis.call('ZADD','KEYS[1]',ARGV[1],ARGV[2])" 1 keyset   123  u123
    
    
    
    def split = body.split(",")
    
    
    
    def type = split[3]
    if (type.substring(0,1) == '1') {
        split[2]=split[2]+"@Start";
    } else {
        split[2]=split[2]+"@End";
    }
    
    Gson gson = new Gson();
    
    Map full= new HashMap();
    full.put("script","return redis.call('ZADD',KEYS[2],KEYS[1],KEYS[3])");
    full.put("args",new ArrayList());
    full.put("keys",split);
    
    String json = gson.toJson(full);
    println json
    
    Map m=gson.fromJson(json, HashMap.class);
    println m
    
    
    //StringBuffer sb = new StringBuffer("return redis.call('ZADD','");
    //
    //sb.append(split[1]).append("',").append(split[0]).append(",'").append(split[2]);
    //if (type.substring(0,1) == '1') {
    //    sb.append("@Start'");
    //} else {
    //    sb.append("@End'");
    //}
    //sb.append(")");
    //println sb
    
    def resultMap = [:]
    
    resultMap["head"] = head
    resultMap["body"] = json
    
    return resultMap

```

CHANGELOG.md
# key-switch
2017-06-26
    分卷脚本 timestamp 加入到 head
    g-netlogact-timestamp-search-replace.groovy
    
2017-06-24

1.^A 的实际编码为\001 ；如何把文件中的tab分隔符CRTL+A,也就是\001,换成 tab分隔符?

    PIG中输入输出分隔符默认是制表符\t，而到了hive中，默认变成了八进制的\001， 也就是ASCII： 
    
    sed 's/\o001/\t/g' yourfile 
    tr '\001' '\t' < yourfile
    
    

2017-06-23

增加 gson 支持

    search-replace 支持转换为 json 数据格式；
    search-replace 支持转换为 redis-lua 脚本，支撑 flume-redis 通用组件。
    
在凌晨1点-4点进行数据清理;保留4天的数据
    
        
    StringBuffer sb1 = new StringBuffer("redis.call('ZADD',KEYS[2],KEYS[1],KEYS[3]);");
    
    //println curhour
    //凌晨进行数据清理 0-4点
    String curhour = new Date().format('HH')
    if (curhour.toInteger() >= 0 && curhour.toInteger() <= 4) {
    
        sb1.append("redis.call('ZREMRANGEBYSCORE',KEYS[2],0,KEYS[4])");
    
        Calendar date = Calendar.getInstance();
        date.set(Calendar.DATE, date.get(Calendar.DATE) - 7);
        println date.format('yyyyMMddHHmmss')
        split = Arrays.copyOf(split, split.length + 1);
        split[4] = date.format('yyyyMMddHHmmss');
    }
   
    

2017-06-21
    增加 redis pool 调用示例
    在 IDE 直接运行脚本测试,再到 Flume 进行联调；
    完成flume-groovy-redis联调；
    
2017-06-20
    加入 redis 支持，便于进行规则计算。
    可以整合 RocksDB涉及到 key 的计算；嵌入式，外部数据更新走 web 接口。todo
    
2016-01-03
    更改项目名称；
    
    
2016-12-16 完成flume 正则拦截器的替换

Search and Replace Interceptor；--> RuleSearchAndReplaceInterceptor$Builder
Regex Filtering Interceptor；  -->RuleFilteringInterceptor$Builder
Regex Extractor Interceptor；-->RuleSearchAndReplaceInterceptor$Builder
---------------------filter dsl
println head
println body

return true

---------------------search replace dsl
println head
println body
body=body.replace('a','aaa')
head["newhead"]='abcd'

def resultMap = [:]

resultMap["head"]=head
resultMap["body"]=body

return resultMap



a1.sources.r1.interceptors = i1 i2 i3
a1.sources.r1.interceptors.i1.type = regex_extractor
a1.sources.r1.interceptors.i1.regex = (.+):(.+):(.+):(.+)
a1.sources.r1.interceptors.i1.serializers = s1 s2 s3 s4
a1.sources.r1.interceptors.i1.serializers.s1.name = routing-key
a1.sources.r1.interceptors.i1.serializers.s2.name = starttime
a1.sources.r1.interceptors.i1.serializers.s3.name = endtime
a1.sources.r1.interceptors.i1.serializers.s4.name = systime

#如果excludeEvents设为true，则表示过滤掉以lxw1234开头的events。
#在使用 Regex Filtering Interceptor的时候一个属性是excludeEvents
#当它的值为true 的时候，过滤掉匹配到当前正则表达式的一行
#当它的值为false的时候，就接受匹配到正则表达式的一行

#a1.sources.r1.interceptors = i2
a1.sources.r1.interceptors.i2.type = com.supermy.flume.interceptor.RuleFilteringInterceptor$Builder
#通过脚本的返回值，数据是否向下传递 参数 body / head(json 格式) 可以引用
a1.sources.r1.interceptors.i2.rule = /etc/flume/conf/g-filter.groovy
a1.sources.r1.interceptors.i2.ruleName = filterGroovy

a1.sources.r1.interceptors.i3.type = com.supermy.flume.interceptor.RuleSearchAndReplaceInterceptor$Builder
#通过脚本的返回值，数据是否向下传递 参数 body / head(json 格式) 可以引用
a1.sources.r1.interceptors.i3.searchReplaceDsl = /etc/flume/conf/g-search-replace.groovy
a1.sources.r1.interceptors.i3.searchReplaceKey = searchReplaceGroovy


2016-12-14 完成拦截器项目的初始化
自定义拦截器，通用的groovy 脚本语言作为规则语音。


src/main/java/com/supermy/flume/interceptor/RuleFilteringInterceptor.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.supermy.flume.interceptor;


import com.google.common.collect.Lists;
import groovy.lang.Binding;
import org.apache.flume.Context;
import org.apache.flume.Event;
import org.apache.flume.interceptor.Interceptor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.util.List;

import static com.supermy.flume.interceptor.RuleFilteringInterceptor.Constants.*;

/**
 * Interceptor that filters events selectively based on a configured regular
 * expression matching against the event body.
 *
 * This supports either include- or exclude-based filtering. A given
 * interceptor can only perform one of these functions, but multiple
 * interceptor can be chained together to create more complex
 * inclusion/exclusion patterns. If include-based filtering is configured, then
 * all events matching the supplied regular expression will be passed through
 * and all events not matching will be ignored. If exclude-based filtering is
 * configured, than all events matching will be ignored, and all other events
 * will pass through.
 *
 * Note that all regular expression matching occurs through Java's built in
 * java.util.regex package.
 *
 * Properties:<p>
 *
 *   regex: Regular expression for matching excluded events.
 *          (default is ".*")<p>
 *
 *   excludeEvents: If true, a regex match determines events to exclude,
 *                  otherwise a regex determines events to include
 *                  (default is false)<p>
 *
 * Sample config:<p>
 *
 * <code>
 *   agent.sources.r1.channels = c1<p>
 *   agent.sources.r1.type = SEQ<p>
 *   agent.sources.r1.interceptors = i1<p>
 *   agent.sources.r1.interceptors.i1.type = REGEX<p>
 *   agent.sources.r1.interceptors.i1.regex = (WARNING)|(ERROR)|(FATAL)<p>
 * </code>
 *
 */

/**
 * @author james mo
 * 利用脚本作为规则判定；
 *
 * Properties:<p>
 *
 *   rulename:
 *   rule: groovy 脚本进行匹配 Regular expression for matching excluded events.
 *          (default is ".*")<p>
 *
 *   excludeEvents: If true, a regex match determines events to exclude,
 *                  otherwise a regex determines events to include
 *                  (default is false)<p>
 *
 *
 * <code>
 *   agent.sources.r1.channels = c1<p>
 *   agent.sources.r1.type = SEQ<p>
 *   agent.sources.r1.interceptors = i1<p>
 *   agent.sources.r1.interceptors.i1.type = RULE<p>
 *   agent.sources.r1.interceptors.i1.rule = (WARNING)|(ERROR)|(FATAL)<p>
 * </code>
 *
 */
public class RuleFilteringInterceptor implements Interceptor {

    private static final Logger logger = LoggerFactory
            .getLogger(RuleFilteringInterceptor.class);

    private final String rule; //groovy file
    private final String rulename;

    private final boolean excludeEvents;

    /**
     * Only {@link RuleFilteringInterceptor.Builder} can build me
     */
    private RuleFilteringInterceptor(String rule, String rulename, boolean excludeEvents) {

        this.rule = rule;
        this.rulename = rulename;
        this.excludeEvents = excludeEvents;
    }

    @Override
    public void initialize() {
        // no-op
    }


    @Override
    /**
     * Returns the event if it passes the regular expression filter and null
     * otherwise.
     *
     * 单个event拦截逻辑
     *
     */
    public Event intercept(Event event) {
        // We've already ensured here that at most one of includeRegex and
        // excludeRegex are defined.
        // 调用脚本规则进行判定 groovy

        //输入参数
        Binding binding = new Binding();
        binding.setVariable("body", new String(event.getBody()));
        binding.setVariable("head", event.getHeaders());

        logger.debug(String.format(
                "flume 的值:  body=%s,head=%s",
                new String(event.getBody()), event.getHeaders()));

        logger.debug(String.format(
                "flume 配置参数:  rule=%s,rulename=%s",
                rule, rulename));

        File f =new File(rule);
        boolean result = (Boolean) GroovyShellJsonExample.getShell(rulename+f.lastModified(), f, binding);

        if (!excludeEvents) {
            //if (regex.matcher(new String(event.getBody())).find()) {// TODO: 16/12/14
            if (result) {
                return event;
            } else {
                return null;
            }
        } else {
            //if (regex.matcher(new String(event.getBody())).find()) {// TODO: 16/12/14
            if (result) {// TODO: 16/12/14
                return null;
            } else {
                return event;
            }
        }
    }

    /**
     * Returns the set of events which pass filters, according to
     * 批量event拦截逻辑
     * {@link #intercept(Event)}.
     * @param events
     * @return
     */
    @Override
    public List<Event> intercept(List<Event> events) {
        List<Event> out = Lists.newArrayList();
        for (Event event : events) {
            Event outEvent = intercept(event);
            if (outEvent != null) {
                out.add(outEvent);
            }
        }
        return out;
    }

    @Override
    public void close() {
        // no-op
    }

    /**
     * Builder which builds new instance of the StaticInterceptor.
     * 相当于自定义Interceptor的工厂类
     * 在flume采集配置文件中通过制定该Builder来创建Interceptor对象
     * 可以在Builder中获取、解析flume采集配置文件中的拦截器Interceptor的自定义参数：
     * 字段分隔符，字段下标，下标分隔符、加密字段下标 ...等
     *
     */
    public static class Builder implements Interceptor.Builder {

        private String rule;
        private String rulename;
        private boolean excludeEvents;

        @Override
        public void configure(Context context) {
            String ruleString = context.getString(RULE, DEFAULT_RULE);
            String ruleNameString = context.getString(RULE_NAME, DEFAUNLT_RULE_NAME);

            //rule = Pattern.compile(ruleString); //// TODO: 16/12/14  groovy 脚本替换
            rule = ruleString;
            rulename = ruleNameString;

            excludeEvents = context.getBoolean(EXCLUDE_EVENTS,
                    DEFAULT_EXCLUDE_EVENTS);
        }

        @Override
        public Interceptor build() {
            logger.info(String.format(
                    "Creating RegexFilteringInterceptor: rule=%s,rulename=%s,excludeEvents=%s",
                    rule, rulename, excludeEvents));
            return new RuleFilteringInterceptor(rule, rulename, excludeEvents);
        }
    }

    /**
     * 常量
     */
    public static class Constants {

        public static final String RULE = "rule";
        public static final String RULE_NAME = "ruleName";

        public static final String DEFAULT_RULE = ".*";
        public static final String DEFAUNLT_RULE_NAME = "filterRule";


        public static final String EXCLUDE_EVENTS = "excludeEvents";
        public static final boolean DEFAULT_EXCLUDE_EVENTS = false;
    }

}